go/src/lib.rs

```diff
@@ -1,9 +1,162 @@
+///
+/// The returned `ConnectionResponse` should be manually freed by calling `free_connection_response`, otherwise a memory leak will occur. It should be freed whether or not an error occurs.
+///
+/// `connection_request_bytes` is an array of bytes that will be parsed into a Protobuf `ConnectionRequest` object.
+/// `connection_request_len` is the number of bytes in `connection_request_bytes`.
+/// `success_callback` is the callback that will be called in the case that the Redis command succeeds.
+/// `failure_callback` is the callback that will be called in the case that the Redis command fails.
+///
+/// # Safety
+///
+/// * `connection_request_bytes` must point to `connection_request_len` consecutive properly initialized bytes. It must be a well-formed Protobuf `ConnectionRequest` object. The array must be allocated by the caller and subsequently freed by the caller after this function returns.
+/// * `connection_request_len` must not be greater than the length of the connection request bytes array. It must also not be greater than the max value of a signed pointer-sized integer.
+/// * The `conn_ptr` pointer in the returned `ConnectionResponse` must live until it is passed into `close_client`.
+/// * The `connection_error_message` pointer in the returned `ConnectionResponse` must live until the returned `ConnectionResponse` pointer is passed to `free_connection_response`.
+/// * Both the `success_callback` and `failure_callback` function pointers need to live until the client is closed via `close_client` since they are used when issuing Redis commands.
+// TODO: Consider making this async
+#[no_mangle]
+pub unsafe extern "C" fn create_client(
+    connection_request_bytes: *const u8,
+    connection_request_len: usize,
+    success_callback: SuccessCallback,
+    failure_callback: FailureCallback,
+) -> *const ConnectionResponse {
+    let request_bytes =
+        unsafe { std::slice::from_raw_parts(connection_request_bytes, connection_request_len) };
+    let response = match create_client_internal(request_bytes, success_callback, failure_callback) {
+        Err(err) => ConnectionResponse {
+            conn_ptr: std::ptr::null(),
+            connection_error_message: CString::into_raw(CString::new(err).unwrap()),
+        },
+        Ok(client) => ConnectionResponse {
+            conn_ptr: Box::into_raw(Box::new(client)) as *const c_void,
+            connection_error_message: std::ptr::null(),
+        },
+    };
+    Box::into_raw(Box::new(response))
+}
+
+/// Closes the given client, deallocating it from the heap.
+///
+/// `client_ptr` is a pointer to the client returned in the `ConnectionResponse` from `create_client`.
+///
+/// # Safety
+///
+/// * `client_ptr` must be obtained from the `ConnectionResponse` returned from `create_client`.
```

Comment:
I'm not sure why you're including this detail. I should be a valid client, but why do we specifically have to use create_client to initialize the client?  In the future, we could have multiple client initialization APIs, and they may not be called create_client.

Reply:
Suggest removing this line.

Reply:
I thought I added this because of another comment stating this should be added, but I can't seem to find it now. I can remove it again though, if you really want me to.

Reply:
+1 for removing or reword. the ptr could be either null

Reply:
@Yury-Fridlyand We don't have to document that the pointer must not be null now because we now have the explicit check for a null pointer. Passing a null pointer is safe now.

Reply:
you need to clarify that `close_client` can only be called once per client. calling it twice is undefined bahavior, since the address will be freed twice.